src/Compilers/Core/Portable/Compilation/IExpression.cs

```diff
@@ -0,0 +1,776 @@
+    {
+        IPropertySymbol Property { get; }
+    }
+
+    public interface IConditionalAccess : IExpression
+    {
+        IExpression Access { get; }
+    }
+
+    /// <summary>
+    /// Represents a unary, binary, relational, or conversion operation that can use an operator method.
+    /// </summary>
+    public interface IWithOperator : IExpression
+    {
+        /// <summary>
+        /// True if and only if the operation is performed by an operator method.
+        /// </summary>
+        bool UsesOperatorMethod { get; }
+        /// <summary>
+        /// Operation method used by the operation, null if the operation does not use an operator method.
+        /// </summary>
+        IMethodSymbol Operator { get; }
+    }
+
+    /// <summary>
+    /// Represents an operation with one operand.
+    /// </summary>
+    public interface IUnary : IWithOperator
+    {
+        UnaryOperationKind UnaryKind { get; }
+        IExpression Operand { get; }
+    }
+
+    /// <summary>
+    /// Kinds of unary operations.
+    /// </summary>
+    public enum UnaryOperationKind
+    {
+        None,
+
+        OperatorBitwiseNegation,
+        OperatorLogicalNot,
+        OperatorPostfixIncrement,
+        OperatorPostfixDecrement,
+        OperatorPrefixIncrement,
+        OperatorPrefixDecrement,
+        OperatorPlus,
+        OperatorMinus,
+        OperatorTrue,
+        OperatorFalse,
+
+        IntegerBitwiseNegation,
+        IntegerPlus,
+        IntegerMinus,
+        IntegerPostfixIncrement,
+        IntegerPostfixDecrement,
+        IntegerPrefixIncrement,
+        IntegerPrefixDecrement,
+
+        UnsignedPostfixIncrement,
+        UnsignedPostfixDecrement,
+        UnsignedPrefixIncrement,
+        UnsignedPrefixDecrement,
+
+        FloatingPlus,
+        FloatingMinus,
+        FloatingPostfixIncrement,
+        FloatingPostfixDecrement,
+        FloatingPrefixIncrement,
+        FloatingPrefixDecrement,
+
+        DecimalPlus,
+        DecimalMinus,
+        DecimalPostfixIncrement,
+        DecimalPostfixDecrement,
+        DecimalPrefixIncrement,
+        DecimalPrefixDecrement,
+
+        BooleanBitwiseNegation,
+        BooleanLogicalNot,
+
+        EnumPostfixIncrement,
+        EnumPostfixDecrement,
+        EnumPrefixIncrement,
+        EnumPrefixDecrement,
+
+        PointerPostfixIncrement,
+        PointerPostfixDecrement,
+        PointerPrefixIncrement,
+        PointerPrefixDecrement,
+
+        DynamicBitwiseNegation,
+        DynamicLogicalNot,
+        DynamicTrue,
+        DynamicFalse,
+        DynamicPlus,
+        DynamicMinus,
+        DynamicPostfixIncrement,
+        DynamicPostfixDecrement,
+        DynamicPrefixIncrement,
+        DynamicPrefixDecrement,
+
+        ObjectPlus,
+        ObjectMinus,
+        ObjectNot
+    }
+
+    /// <summary>
+    /// Represents an operation with two operands that produces a result with the same type as at least one of the operands.
+    /// </summary>
```

Comment:
This definitely seems odd to me.  The distinction between IRelational and IBinary seems very thin.  The two interfaces look like this:

```c#
+    public interface IBinary : IWithOperator
+    {
+        BinaryOperationKind BinaryKind { get; }
+        IExpression Left { get; }
+        IExpression Right { get; }
+    }

+    public interface IRelational : IWithOperator
+    {
+        RelationalOperationKind RelationalKind { get; }
+        IExpression Left { get; }
+        IExpression Right { get; }
+    }
```

Should we just have IBinaryExpression and BinaryExpressionKind?